### PR TITLE
Feat: Full backup/restore engine with P2P sharing via Hypercore

### PR DESCRIPTION
**Pull request by Ayush-Vish**
## Related Issue (if any)

GSoC 2026  - Phases 1 & 2 (Core Backup Engine + P2P Sharing)

### Describe the add-ons or changes you've made

Implements the complete backup and restore system for Peersky Browser. Users can create portable `.zip` backups of their browser state (tabs, P2P data, chat rooms, ENS cache), restore them locally, and share backups over IPFS or Hyperdrive using content addresses.

**Core Backup Engine** (`src/backup/`)
- `backup-core.js` -- zip creation/extraction with SHA-256 manifest, zip-slip protection, lock file exclusion, atomic `.tmp` write
- `backup-worker.js` -- offloads heavy I/O to a worker thread so the UI stays responsive
- `backup-manager.js` -- orchestrates create/restore with quiesce-then-zip strategy (suspends IPFS + Hyper before snapshotting, resumes after)
- Transactional restore: extract to temp dir, verify checksums, then overwrite userData
- Orphaned temp directory cleanup on boot

**P2P Sharing** (`src/backup/`)
- `p2p-backup.js` -- publish backups to IPFS (`addByteStream`) or Hyperdrive; download from `ipfs://` CID or `hyper://` address
- `ipc.js` -- 6 IPC handlers: `backup-create`, `backup-validate`, `backup-restore`, `backup-upload`, `backup-restore-cid`, `backup-relaunch`
- Cache wiring: uploads/downloads logged into `ipfsCache.json` / `hyperCache.json`

**Protocol Enhancements** (`src/protocols/`)
- `ipfs-handler.js` -- `suspendIPFS()`, `resumeIPFS()`, `ipfsPublishFile()`, `ipfsFetchToFile()`; `provideCidWithRetry()` utility; periodic DHT re-provide (6h)
- `hyper-handler.js` -- `suspendHyper()`, `resumeHyper()`, `hyperPublishFile()`, `hyperFetchToFile()` with retry + backoff; ephemeral ports to avoid bind conflicts
- `helia.js` -- ephemeral listen ports (`/tcp/0`, `/udp/0`)

**Backup UI**
- `backup.html` + `backup.js` -- full page at `peersky://backup`
- Settings sidebar link
- Onboarding restore support (`onboarding-restore-zip`, `onboarding-restore-cid`)

**Documentation**
- `docs/Backup-Restore.md`

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Code style update (formatting, local variables)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update

## How Has This Been Tested?

**27 unit tests passing** across 3 test suites (`npm run test:backup`):

| Suite | Tests |
|---|---|
| `backup-core` (5) | zip round-trip + checksum verification, lock file exclusion, zip-slip rejection, checksum mismatch detection, invalid manifest rejection |
| `backup-manager` (3) | P2P services stay stopped after successful restore, services resume on failure, worker error propagation |
| `p2p-backup` (19) | `parseIpfsAddress` edge cases (x8), `uploadBackup` IPFS + Hyper with cache wiring (x2), `downloadBackupFromAddress` IPFS/Hyper paths with cache wiring + status forwarding (x7), plus 2 existing `ipfs-handler` tests for CID provide |

All existing test suites continue to pass (`test:p2p`, `test:extensions`, `test:security`).

## Checklist:

- [x] My code follows the "contribution guidelines" of this project.
- [x] I have performed a self-review of my own code.
- [x] I have commented my code, particularly wherever it was hard to understand.
- [x] My changes generate no new warnings.
- [ ] Any dependent changes have been merged and published in downstream modules.

### CODE REVIEW

**Comment by Ayush-Vish** [comment]
# Backup/ Restore using local zip 

1. A fresh instance is left, and an already created instance on the right
Screenshot from 2026-06-24 00-27-07

2. go to peersky://backup/ page and create a backup
Screenshot from 2026-06-24 00-27-38

3. Restore the create backup using the upload zip button
Screenshot from 2026-06-24 00-28-01

4. Backup Created 
Screenshot from 2026-06-24 00-28-57

File: docs/Backup.md
Comment (akhileshthite) [inline docs/Backup.md:1]: Rename this to `Backup`

File: src/protocols/hyper-handler.js [hunk 1]
```diff
@@ -110,6 +113,112 @@ async function initializeHyperSDK (options) {
...
+  const f = await initializeHyperSDK()
+  const driveName = `peersky-backup-${Date.now()}`
+
+  // Resolve the drive's public key (hypercore-fetch returns it for ?key=).
+  // Use hyper://localhost/ as the host, which is the required format.
+  const keyResp = await f(`hyper://localhost/?key=${driveName}`, { method: 'POST' })
+  const keyText = await keyResp.text()
+  log.info(`hyperPublishFile ?key= status: ${keyResp.status}, text: ${keyText}`)
+
+  const match = keyText.match(/([0-9a-zA-Z]{52,64})/)
```
Comment (akhileshthite) [inline src/protocols/hyper-handler.js]: This feels brittle; if `hypercore-fetch` changes its response format in a future update, this silently breaks.

File: src/protocols/hyper-handler.js [hunk 2]
```diff
@@ -110,6 +113,112 @@ async function initializeHyperSDK (options) {
...
+    body,
+    duplex: 'half'
+  })
+  if (!putResp.ok) throw new Error(`Hyperdrive write failed: ${putResp.status}`)
+
+  log.info(`Backup published to Hyperdrive: ${driveKey}`)
+  return { key: driveKey, fileName, address: `hyper://${driveKey}/${fileName}` }
+}
+
+async function waitForDriveReady (url) {
```
Comment (akhileshthite) [inline src/protocols/hyper-handler.js:225]: `waitForDriveReady` is now called on every hyper request (see line 318), not just backup fetches. For regular browsing of known drives, this adds unnecessary peer-discovery latency on each navigation. Worth gating this behind a check like `if (drive.core.length === 0)` at the call site.

File: src/protocols/hyper-handler.js [hunk 3]
```diff
@@ -110,6 +113,112 @@ async function initializeHyperSDK (options) {
...
+    }
+    await drive.update()
+    log.info(`Finished waiting for peers for ${hostname}, core length is now ${drive.core.length}`)
+  } catch (err) {
+    log.error(`Error waiting for peers for ${url}:`, err)
+  }
+}
+
+// Stream a hyper:// file address to destPath on disk.
+export async function hyperFetchToFile (address, destPath, onStatus) {
```
Comment (akhileshthite) [inline src/protocols/hyper-handler.js:248]: The retry loop always fires a new `f(address)` request but never re-calls `waitForDriveReady` between attempts. If the first attempt fails because no peers were found yet, subsequent attempts will likely fail for the same reason. Might be worth calling `waitForDriveReady(address)` inside the loop too.

File: src/protocols/hyper-handler.js [hunk 4]
```diff
@@ -110,6 +113,112 @@ async function initializeHyperSDK (options) {
   return fetch
 }
 
+// Close the corestore entirely so its RocksDB state is strictly frozen on disk.
+export async function suspendHyper () {
```
Comment (akhileshthite) [inline src/protocols/hyper-handler.js:180]: Let's move all of this to `backup/` directory in a file called `hyper-backup.js` and then import here.

File: src/protocols/ipfs-handler.js [hunk 5]
```diff
@@ -18,6 +18,105 @@ import { enforceExtensionWritePolicy } from '../extensions/request-policy.js'
...
 const log = createLogger('protocols:ipfs')
 
+// Shared references to the running Helia node, set by the first createHandler.
+// Used by the backup feature to publish/fetch content without a second node.
+let sharedNode = null
+let sharedUnixFs = null
+
+async function provideCidWithRetry (node, cid, options = {}) {
+  const {
+    maxAttempts = 3,
```
Comment (akhileshthite) [inline src/protocols/ipfs-handler.js]: Let's move all of this to `backup/` directory in a file called `ipfs-backup.js` and then import here.

File: src/protocols/ipfs-handler.js [hunk 6]
```diff
@@ -18,6 +18,105 @@ import { enforceExtensionWritePolicy } from '../extensions/request-policy.js'
 
 const log = createLogger('protocols:ipfs')
 
+// Shared references to the running Helia node, set by the first createHandler.
+// Used by the backup feature to publish/fetch content without a second node.
+let sharedNode = null
+let sharedUnixFs = null
```
Comment (akhileshthite) [inline src/protocols/ipfs-handler.js:22]: sharedNode and sharedUnixFs are module-level let bindings set inside createHandler. If createHandler is somehow called twice (say during a test or a hot-reload path), the old references are silently overwritten. A guard or at least a log warning would be nice here.

File: src/protocols/ipfs-handler.js [hunk 7]
```diff
@@ -18,6 +18,105 @@ import { enforceExtensionWritePolicy } from '../extensions/request-policy.js'
...
+    await sharedNode.stop()
+  }
+}
+
+export async function resumeIPFS () {
+  if (sharedNode) {
+    log.info('Resuming IPFS node after backup...')
+    await sharedNode.start()
+  }
+}
```
Comment (akhileshthite) [inline src/protocols/ipfs-handler.js:33]: `suspendIPFS` calls `sharedNode.stop()` but doesn't null out `sharedNode`; so `resumeIPFS` can call `.start()` on the same instance. That's fine, but `ipfsPublishFile` checks `if (!sharedNode)` to know if IPFS is ready. After `suspendIPFS`, `sharedNode` is still truthy even though the node is stopped. If anything calls `ipfsPublishFile` between suspend and resume it'll try to use a stopped node. Maybe set a `suspended = true` flag and check it?

File: src/backup/backup-core.js [hunk 8]
```diff
@@ -0,0 +1,258 @@
...
+async function sha256DirOld (dir) {
+  const files = (await listDirFilesOld(dir)).sort()
+  const hash = crypto.createHash('sha256')
+  for (const rel of files) {
+    hash.update(rel)
+    hash.update(await sha256File(path.join(dir, rel)))
+  }
+  return hash.digest('hex')
+}
+
```
Comment (akhileshthite) [inline src/backup/backup-core.js:136]: Can we add a comment with when this can be removed (e.g., "Safe to remove after version X when all users have re-created backups")? Otherwise this dead-looking code will confuse future contributors.

File: src/backup/backup-core.js [hunk 9]
```diff
@@ -0,0 +1,258 @@
...
+  { name: 'peersky-chat-rooms.json', type: 'file' },
+  { name: 'peersky-ports.json', type: 'file' },
+  { name: 'ipfs', type: 'dir' },
+  { name: 'hyper', type: 'dir' }
+]
+
+// Live database lock and in-flight files that must never enter the bundle.
+// The hypercore-storage device file (CORESTORE) is intentionally left in so
+// manifest hashes stay stable across versions; it is dropped on restore
+// instead, since its recorded inode/xattr never survive a copy anyway.
```
Comment (akhileshthite) [inline src/backup/backup-core.js]: The comment says "CORESTORE is intentionally left in so manifest hashes stay stable" but then it's "dropped on restore instead." I don't see where in the restore path CORESTORE is explicitly removed. Is that handled in `backup-manager.js`?

Comment (akhileshthite) [reply src/backup/backup-core.js]: Also let's just reduce the comment length to two lines max everywhere in this PR.

File: src/backup/backup-manager.js
Comment (akhileshthite) [inline src/backup/backup-manager.js:1]: The test shows that on successful restore, P2P services are NOT resumed. This means the app is in a degraded state until the mandatory restart. That's fine, but if the user dismisses the restart prompt or the relaunch fails, they'll have a broken session with no IPFS/Hyper. Worth adding a note in the UI that the browser must be restarted, or force-quit after a timeout.

File: src/window-manager.js [hunk 10]
```diff
@@ -537,6 +542,11 @@ class WindowManager {
   }
 
   async saveCompleteState () {
+    if (this.skipSaveOnQuit) {
+      log.info('Skipping saveCompleteState because skipSaveOnQuit is true.')
+      return
+    }
+
```
Comment (akhileshthite) [inline src/window-manager.js:554]: `skipSaveOnQuit` prevents all state saving, not just backup-related saves. If the app crashes between setting this flag and actually quitting, the user loses their session entirely. Is this flag only set immediately before `app.quit()`?  
just want to make sure there's no window where a crash leaves this sticky.

File: package.json [hunk 11]
```diff
@@ -214,6 +216,11 @@
     "sinon": "^21.0.2",
     "standard": "^17.1.2"
   },
+  "overrides": {
+    "hypercore": "^11.33.1",
+    "corestore": "^7.10.1",
+    "hypercore-storage": "^3.1.1"
+  },
```
Comment (akhileshthite) [inline package.json]: Pinning hypercore, corestore, and hypercore-storage via overrides is a pretty aggressive move. This forces all transitive consumers to use these versions, which could cause subtle breakage in other deps that depend on older APIs. Please remove.

File: src/settings-manager.js [hunk 12]
```diff
@@ -532,6 +533,11 @@ class SettingsManager {
       // Start with defaults
       this.settings = { ...DEFAULT_SETTINGS }
 
+      // Existing profiles upgrading to this version should bypass onboarding
+      if (loaded && Object.keys(loaded).length > 0 && !('onboardingCompleted' in loaded)) {
+        this.settings.onboardingCompleted = true
+      }
```
Comment (akhileshthite) [inline src/settings-manager.js:545]: This silently sets onboarding as completed for existing users. Makes sense, but it means if a user's settings file gets corrupted to `{} (empty object)`, they'd see onboarding again. Just another edge case..

File: test/integration/extension-restart-harness.mjs [hunk 13]
```diff
@@ -24,6 +24,15 @@ function emitResult (payload) {
...
 }
 
+async function safeShutdownAndExit (code) {
+  try {
+    await withTimeout(extensionManager.shutdown(), 5000, 'shutdown')
+  } catch {
+  }
+  app.exit(code)
+  setTimeout(() => process.exit(code), 3000)
+}
```
Comment (akhileshthite) [inline test/integration/extension-restart-harness.mjs:40]: suggests `app.exit()` sometimes doesn't actually kill the process in CI. Is this a known Electron quirk? If so, this is a reasonable workaround; but might be worth adding a comment explaining why it's needed.

File: src/protocols/helia/helia.js [hunk 14]
```diff
@@ -58,11 +58,11 @@ export async function createNode() {
...
         '/ip4/0.0.0.0/tcp/0',
-        '/ip4/0.0.0.0/tcp/4002/ws',
-        '/ip4/0.0.0.0/udp/4003/webrtc-direct',
+        '/ip4/0.0.0.0/tcp/0/ws',
+        '/ip4/0.0.0.0/udp/0/webrtc-direct',
         '/ip6/::/tcp/0',
-        '/ip6/::/tcp/4002/ws',
-        '/ip6/::/udp/4003/webrtc-direct',
+        '/ip6/::/tcp/0/ws',
+        '/ip6/::/udp/0/webrtc-direct',
```
Comment (akhileshthite) [inline src/protocols/helia/helia.js:65]: Why the ports are changed? Please revert.

**Review by akhileshthite — CHANGES_REQUESTED** [review]
Good start!

File: src/backup/hyper-backup.js [hunk 15]
```diff
@@ -0,0 +1,99 @@
...
+    body,
+    duplex: 'half'
+  })
+  if (!putResp.ok) throw new Error(`Hyperdrive write failed: ${putResp.status}`)
+
+  // Flush swarm announcement so the drive is discoverable before returning
+  if (sdk) {
+    try {
+      const drive = await sdk.getDrive(`hyper://${driveKey}/`)
+      sdk.joinCore(drive.core)
```
Comment (HarshalAtre) [inline src/backup/hyper-backup.js]: `joinCore()` is asynchronous and must be awaited here. Without `await`, the transfer QR can be returned before the backup drive finishes announcing, causing the initial mobile fetch to fail with `404 Peers Not Found`. 

so make it `await sdk.joinCore(drive.core)`

File: src/backup/backup-core.js [hunk 16]
```diff
@@ -0,0 +1,285 @@
...
+// Top-level entries backed up and restored. Files are optional;
+// directories are streamed whole minus the files listed in SKIP_NAMES.
+export const STANDARD_BACKUP_TARGETS = [
+  { name: 'lastOpened.json', type: 'file' },
+  { name: 'tabs.json', type: 'file' },
+  { name: 'ensCache.json', type: 'file' },
+  { name: 'ipfsCache.json', type: 'file' },
+  { name: 'hyperCache.json', type: 'file' },
+  { name: 'ipfs', type: 'dir' },
+  { name: 'hyper', type: 'dir' }
```
Comment (akhileshthite) [inline src/backup/backup-core.js]: This is the main blocker for me. `ipfs/` holds `libp2p-key` which is the libp2p private key, and `hyper/db` is the corestore that holds secret keys for writable cores. So a standard backup carries the full crypto identity even though we exclude the json files.

I checked this by building a real zip:
entries in a STANDARD (non-identity) backup zip
```
manifest.json / tabs.json / ipfs/libp2p-key / hyper/db/corestore-secret
libp2p private key present in standard backup: true
peersky-ports.json present: false
```

This matters because `uploadBackup` publishes the same zip with no encryption. Please either strip the key material from the standard target list, or encrypt before any upload. Right now the doc note saying identity files are excluded is not true in practice.

File: src/backup/p2p-backup.js [hunk 17]
```diff
@@ -0,0 +1,117 @@
...
+    throw new Error('Only raw CIDs or ipfs:// links are supported')
+  }
+  value = value.replace(/^\/+/, '').split('/')[0].split('?')[0]
+  if (!value) throw new Error('Could not read a CID from the input')
+  return value
+}
+
+// Publish a backup zip to a P2P network and return its content address.
+// protocol: 'ipfs' (default) returns a CID, 'hyper' returns a hyper:// URL.
+export async function uploadBackup (zipPath, protocol = 'ipfs') {
```
Comment (akhileshthite) [inline src/backup/p2p-backup.js]: `uploadBackup` sends the zip as is, no encryption. Combined with `ipfs/libp2p-key` and the corestore being inside the bundle, this publishes the user private keys to a public network. IPFS content also gets re-provided by other peers, so there is no delete after the fact.

Please gate this so only an encrypted bundle can be uploaded.

File: src/pages/backup.html [hunk 18]
```diff
@@ -0,0 +1,190 @@
...
+            </div>
+            <div class="setting-control">
+              <button type="button" id="backup-create" class="btn btn-primary">Create Backup</button>
+            </div>
+          </div>
+          <div class="setting-row tall-row">
+            <div class="setting-label">
+              Share via P2P
+              <p class="setting-description">Upload a backup .zip to IPFS/Hypercore and get a
+                CID you can share. Anyone can restore it from the onboarding screen
```
Comment (akhileshthite) [inline src/pages/backup.html]: "Anyone can restore it" is written like a feature but the zip has the user private keys in it. Also line 57 says the backup bundles "P2P identities, PeerChat rooms" while `docs/Backup.md` line 24 says those are excluded. Both cannot be right. Please make the UI and the doc say the same thing after the key material is handled.

File: src/backup/identity-transfer.js [hunk 19]
```diff
@@ -0,0 +1,281 @@
...
+  const time = Math.floor(Date.now() / 60000) // 1 minute window
+  const timeBuffer = Buffer.alloc(8)
+  timeBuffer.writeBigUInt64BE(BigInt(time))
+  const hmac = crypto.createHmac('sha1', secret)
+  hmac.update(timeBuffer)
+  const hash = hmac.digest()
+  const offset = hash[hash.length - 1] & 0xf
+  const code = (hash.readUInt32BE(offset) & 0x7fffffff) % 1000000
+  return String(code).padStart(6, '0')
+}
```
Comment (akhileshthite) [inline src/backup/identity-transfer.js]: This code can never match what mobile shows. Desktop makes a 6 digit time based code from a random `channel` on a 1 minute window. Mobile makes 6 uppercase hex chars:

`sha256(sourceSigningPublicKey || targetEncryptionPublicKey || nonce)` then `.slice(0, 6).toUpperCase()`

Different input, different hash, different alphabet, and ours rotates every 60 seconds while theirs is fixed. So the user can never match them, and they will just learn to ignore the code. That removes the only MITM protection in this flow.

The mobile derivation is the correct one because it binds `sourceSigningPublicKey`. Please use the same formula here and drop the TOTP.

File: src/pages/static/js/backup.js [hunk 20]
```diff
@@ -0,0 +1,380 @@
...
+      cidValue.textContent = res.address
+      const qrImg = document.getElementById('backup-qr-code')
+      if (qrImg && res.qrDataUrl) {
+        qrImg.src = res.qrDataUrl
+        qrImg.style.display = 'block'
+      } else if (qrImg) {
+        qrImg.style.display = 'none'
+      }
+      cidRow.style.display = ''
+      showStatus(`Encrypted identity transfer uploaded to Hyper.\n\nVERIFICATION CODE: ${res.verificationCode}\n\nScan the QR code below with PeerSky Mobile (Settings > Link Device) to restore identity automatically. Ensure the Verification Code matches exactly!`, 'success')
```
Comment (akhileshthite) [inline src/pages/static/js/backup.js]: We tell the user "Ensure the Verification Code matches exactly!" but the code we show here can never equal the one mobile shows. See my comment on `identity-transfer.js`. Please fix the derivation before this string ships, otherwise we are training users to skip the check.

File: src/backup/backup-manager.js [hunk 21]
```diff
@@ -0,0 +1,187 @@
...
+        applyManifest = await extractAndVerifyIdentityPayload(innerZipPath, innerDir)
+        applyDir = innerDir
+      } else {
+        await verifyManifest(tempDir, manifest)
+      }
+
+      for (const name of Object.keys(applyManifest.files)) {
+        const src = path.join(applyDir, name)
+        const target = path.join(dest, name)
+        await fs.rm(target, { recursive: true, force: true }).catch(() => {})
```
Comment (akhileshthite) [inline src/backup/backup-manager.js]: This deletes the live directory and then copies. If the copy fails in the middle, disk full or crash or IO error, the user has no `ipfs/`, no `hyper/` and no identity, and there is no way back. The PR calls this transactional restore with atomic writes but it is neither.

Please copy into a staging dir next to the target first, then swap with rename, and keep the old one until the swap is done.

Also the `.catch(() => {})` on the `rm` hides a failed delete. If the delete fails the copy still runs with `overwrite: true` and old and new repos get merged, which is the exact thing the doc says must not happen.

File: test/backup/identity-transfer.test.js [hunk 22]
```diff
@@ -0,0 +1,89 @@
...
+    const manifest = await readManifest(outPath)
+    const innerZip = path.join(await makeTempDir('peersky-id-inner-'), 'inner.zip')
+    const decrypted = await decryptIdentityTransferZip(target, extracted, manifest, innerZip)
+
+    expect(decrypted.transfer.targetDeviceType).to.equal('mobile')
+    expect(decrypted.verificationCode).to.equal(deriveVerificationCode(decrypted.transfer))
+
+    const payloadDir = await makeTempDir('peersky-id-payload-')
+    const innerManifest = await extractAndVerifyIdentityPayload(innerZip, payloadDir)
+    expect(Object.keys(innerManifest.files)).to.include.members(['peersky-ports.json', 'peersky-devices.json'])
```
Comment (akhileshthite) [inline test/backup/identity-transfer.test.js]: This test fails on the branch. `npm run test:backup` gives 28 passing and 1 failing.

```
AssertionError: expected [ 'hyper', ...(2) ] to be a superset of [ 'peersky-ports.json', ...(1) ]
```

The transfer here is `mobile`, and `MOBILE_IDENTITY_BACKUP_TARGETS` does not include `peersky-devices.json`, so the assert is stale. The PR body says 27 tests passing, please recheck.

File: src/backup/device-registry.js [hunk 23]
```diff
@@ -0,0 +1,152 @@
...
+
+  return saveDeviceRegistry(userDataDir, registry, keys.signing.secretKey)
+}
+
+export async function assertIdentityImportAllowed (userDataDir, transfer) {
+  const registry = await loadDeviceRegistry(userDataDir)
+  if (!registry || registry.identityId !== transfer.identityId) return true
+
+  return true
+}
```
Comment (akhileshthite) [inline src/backup/device-registry.js]: This function loads the registry, ignores the result and returns true in both paths. So there is no import side check at all. The doc says `identity-transfer.js` does limit enforcement but nothing enforces anything.

File: src/pages/backup.html [hunk 24]
```diff
@@ -0,0 +1,190 @@
...
+        </section>
+
+        <section class="settings-section">
+          <div class="setting-row tall-row">
+            <div class="setting-label">
+              Identity transfer
+              <p class="setting-description">Create an encrypted identity transfer for PeerSky Mobile or Desktop. Enter the 64-character encryption key shown on the receiving device (or scan Mobile QR code). Click "Upload Hyper" to generate a transfer QR code for PeerSky Mobile.</p>
+            </div>
+            <div class="setting-control" style="display: flex; gap: 8px; flex-wrap: wrap; justify-content: flex-end; align-items: center;">
+              <select id="identity-device-type" class="cid-input" style="width: auto; margin-bottom: 0; height: 34px; box-sizing: border-box;">
```
Comment (akhileshthite) [inline src/pages/backup.html]: The device type is picked by the sender in a dropdown. So if I pick "Desktop" while pairing a phone I get as many phones as I want, and the one mobile rule is gone. The receiving device never says what it is.

The type should come from the QR payload that the receiving device generates, not from this select.

File: src/backup/device-registry.js [hunk 25]
```diff
@@ -0,0 +1,152 @@
...
+  }
+  if (!registry.devices.mobile) registry.devices.mobile = []
+
+  if (normalizedType === 'desktop') {
+    registry.devices.desktop.push({
+      encryptionPublicKey: targetEncryptionPublicKey,
+      pairedAt: new Date().toISOString()
+    })
+  } else {
+    registry.devices.mobile = [{
```
Comment (akhileshthite) [inline src/backup/device-registry.js]: Replacing the mobile slot only rewrites a local json file. The old phone still has the transferred keys and keeps full access. So one mobile is a label in the UI, not a real limit.

For this to mean anything we need actual revoke, like rotating the room keys on the desktop so the old phone stops working.

File: src/backup/device-registry.js [hunk 26]
```diff
@@ -0,0 +1,152 @@
...
+  const hash = crypto.createHash('sha256')
+  const names = ['peersky-ports.json', 'peersky-chat-rooms.json']
+  let found = false
+
+  for (const name of names) {
+    try {
+      await fs.readFile(path.join(userDataDir, name))
+      hash.update(name)
+      // Do not hash the mutable file contents which would invalidate the identityId on any change
+      // hash.update(data)
```
Comment (akhileshthite) [inline src/backup/device-registry.js]: This hashes only which filenames exist, never the contents. So `identityId` has only 4 possible values in the whole world, and every install that has both files gets the same one. Any comparison against `identityId` is meaningless, including the one on line 110 and in `assertIdentityImportAllowed`.

If the concern was that content changes invalidate the id, then derive it once from a stable seed and store it, do not hash file names.

File: src/backup/backup-manager.js [hunk 27]
```diff
@@ -0,0 +1,187 @@
...
+
+      // If this was an Identity Transfer, assume ownership of the registry.
+      // This allows the restoring device (Desktop or Mobile) to manage the identity going forward.
+      if (identityTransfer) {
+        const slotType = identityTransfer.transfer.targetDeviceType
+        const keys = await getDeviceKeys(dest)
+        const publicInfo = getPublicDeviceInfo(keys)
+        const registry = await loadDeviceRegistry(dest)
+        if (registry) {
+          registry.ownerSigningPublicKey = publicInfo.signingPublicKey
```
Comment (akhileshthite) [inline src/backup/backup-manager.js]: After a desktop to desktop transfer the receiving desktop makes itself the owner and clears its own slot list. Now two desktops each own a registry for the same identity and there is no sync between them. Each one can then pair its own single mobile, so we end up with two phones on one identity.

Also if a user just deletes `peersky-devices.json` the registry comes back empty because `loadDeviceRegistry` returns null on ENOENT, and the limit resets.

File: src/backup/device-registry.js [hunk 28]
```diff
@@ -0,0 +1,152 @@
...
+  return Buffer.from(hypercoreCrypto.sign(message, signingSecretKey)).toString('hex')
+}
+
+export function verifyRegistry (registry) {
+  if (!registry || typeof registry !== 'object' || !registry.signature) return false
+  const message = Buffer.from(canonicalJson(registryBody(registry)))
+  return hypercoreCrypto.verify(
+    message,
+    Buffer.from(registry.signature, 'hex'),
+    Buffer.from(registry.ownerSigningPublicKey, 'hex')
```
Comment (akhileshthite) [inline src/backup/device-registry.js]: The signature is checked against `ownerSigningPublicKey` which is stored inside the same file. Anyone who can write the file can put their own key there and re-sign. So this proves the file is not corrupted, it does not prove it was not tampered with. Fine as a checksum, but please do not treat it as a trust boundary.

File: src/backup/identity-transfer.js [hunk 29]
```diff
@@ -0,0 +1,281 @@
...
+export function verifyOTP (token, transfer) {
+  return generateTOTP(transfer.channel) === token
+}
+
+export function verifyIdentityTransferSignature (transfer) {
+  const message = Buffer.from(canonicalJson(transferBody(transfer)))
+  return hypercoreCrypto.verify(
+    message,
+    Buffer.from(transfer.signature, 'hex'),
+    Buffer.from(transfer.sourceSigningPublicKey, 'hex')
```
Comment (akhileshthite) [inline src/backup/identity-transfer.js]: Same issue as the registry. We verify against `sourceSigningPublicKey` which comes from the transfer itself, so any sender can make a transfer that verifies. The receiver has no pinned idea of which desktop is the real one.

This is exactly why the SAS code has to work. On mobile the SAS binds `sourceSigningPublicKey`, so comparing it is what actually authenticates the desktop. With our code broken there is nothing left.

File: package.json [hunk 30]
```diff
@@ -220,6 +222,7 @@
     "mime-types": "^2.1.35",
     "multiformats": "^13.3.2",
     "parse-torrent": "^11.0.19",
+    "qrcode": "^1.5.4",
     "scoped-fs": "^1.4.1",
     "undici": "^8.9.0",
     "unzipper": "^0.10.14",
```
Comment (akhileshthite) [inline package.json]: bump this to latest

File: src/backup/ipc.js [hunk 31]
```diff
@@ -0,0 +1,204 @@
+import path from 'path'
+import { ipcMain, dialog, app, BrowserWindow } from 'electron'
+import { createLogger } from '../logger.js'
+import backupManager, { defaultBackupName } from './backup-manager.js'
+import { uploadBackup, downloadBackupFromAddress } from './p2p-backup.js'
+import { getDeviceKeys, getPublicDeviceInfo } from './device-keys.js'
+import { loadDeviceRegistry } from './device-registry.js'
+import QRCode from 'qrcode'
```
Comment (akhileshthite) [inline src/backup/ipc.js]: Use the inbuilt QR generater we have, see https://github.com/p2plabsxyz/peersky-browser/blob/6c749020f45420e058e8adfae590ca6170b0020d/src/pages/nav-box.js#L574

File: package.json [hunk 32]
```diff
@@ -220,6 +222,7 @@
     "mime-types": "^2.1.35",
     "multiformats": "^13.3.2",
     "parse-torrent": "^11.0.19",
+    "qrcode": "^1.5.4",
```
Comment (akhileshthite) [inline package.json]: Remove this package https://github.com/p2plabsxyz/peersky-browser/pull/204/changes#r3808893924

File: src/backup/backup-core.js [hunk 33]
```diff
@@ -0,0 +1,285 @@
...
+  { name: 'peersky-devices.json', type: 'file' }
+]
+
+export const MOBILE_IDENTITY_BACKUP_TARGETS = [
+  { name: 'lastOpened.json', type: 'file' },
+  { name: 'tabs.json', type: 'file' },
+  { name: 'hyper', type: 'dir' },
+  { name: 'peersky-chat-rooms.json', type: 'file' },
+  { name: 'peersky-ports.json', type: 'file' }
+]
```
Comment (akhileshthite) [inline src/backup/backup-core.js:34]: The mobile payload includes the whole `hyper/` dir, but the mobile side caps zip reading at 50 MB. A normal corestore will pass that easily, so real transfers will just fail on the phone. Please check the size before upload and give a clear error.

Comment (akhileshthite) [reply src/backup/backup-core.js:34]: See:
Image

File: src/backup/backup-core.js [hunk 34]
```diff
@@ -0,0 +1,285 @@
...
+    files
+  }
+}
+
+// Stream all present targets plus a manifest into a single .zip at outPath.
+// onProgress receives { processedBytes, totalBytes, entries } updates.
+export async function createBackupZip (userDataDir, outPath, options = {}) {
+  const { peerskyVersion = '', isIdentityTransfer = false, targetDeviceType = 'desktop', onProgress } = options
+
+  const manifest = await buildManifest(userDataDir, peerskyVersion, isIdentityTransfer, targetDeviceType)
```
Comment (akhileshthite) [inline src/backup/backup-core.js]: The manifest hashes the files, then the archiver reads the same files again. Even with services suspended, anything that writes in between gives a zip that passes creation but fails `verifyManifest` on restore. So the user gets a backup that looks fine and cannot be restored, and only finds out later.

Better to hash while streaming into the archive so the manifest matches what actually went in.

**Review by akhileshthite — CHANGES_REQUESTED** [review]
@Ayush-Vish Thanks for the work here, the shape of the crypto envelope is right and the zip slip handling is correct. But I am not able to merge this yet. There are some major blockers.

Also, the mobile device limit is not really enforced. The registry does keep one mobile slot and many desktop slots, but nothing enforces it. `assertIdentityImportAllowed` always returns true, the device type is a dropdown the sender picks, and there is no revoke. So an evicted phone keeps working and picking "Desktop" for a phone gives unlimited phones.

File: src/backup/p2p-backup.js (hunk 17 above)
Comment (akhileshthite) [reply src/backup/p2p-backup.js]: - `ipfs/` can just go. `getLibp2pPrivateKey` in `src/protocols/config.js` regenerates the key when the file is missing, and the blockstore is a refetchable cache. Dropping it costs nothing, removes a private key, and saves 507 MB on my profile.

- `hyper/` has to stay. Without the corestore secrets every drive the user authored becomes read-only forever and their `hyper://` URLs die.

So a full backup is always key material, and it should be encrypted with a user passphrase even for a local file, since a plain zip in Downloads ends up synced to iCloud anyway. `encryptFile` and `decryptFile` are already written in `identity-transfer.js`, they just need a scrypt derived key instead of a random one. No new dependency.

My suggestion is to drop "Share via P2P" from this PR. To be clear I am not asking you to remove P2P upload. `backup-identity-upload-hyper` calls the same `uploadBackup` and that one is fine, because the payload is already sealed to the target device key. The only path I want gone is the one that feeds a plain zip into it. Keeping a copy is covered by the local encrypted backup, and moving to your own device is covered by identity transfer. Publishing to IPFS is irreversible and does not add a case those two miss.

One more thing I noticed while reading the hyper path. It creates a new drive under `hyper://localhost/?key=...` each time, which is a writable drive in the user's own corestore. So every share leaves a drive in `hyper/` that never gets cleaned up, and it then rides along in every later backup.

**Review by akhileshthite — DISMISSED** [review]
LGTM, thanks!
